### PR TITLE
bugfix(compatibility): Restore retail compatibility after sequential script change in ScriptEngine

### PR DESCRIPTION
**Pull request by Caball009**
* Relates to https://github.com/TheSuperHackers/GeneralsGameCode/issues/828
* Follow up for #629

#629 introduced a change that is not compatible with retail. This PR fixes that by putting it behind the retail compatibility macro.

## TODO:
- [ ] Replicate in Generals

### CODE REVIEW

File: GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp [hunk 1]
```diff
@@ -7877,12 +7877,36 @@ void ScriptEngine::setSequentialTimer(Team *team, Int frameCount)
 
 void ScriptEngine::evaluateAndProgressAllSequentialScripts( void )
 {
+#if (defined(_MSC_VER) && _MSC_VER < 1300) && RETAIL_COMPATIBLE_CRC
```
Comment (Caball009) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7880]: Would it be preferable to change
`#if (defined(_MSC_VER) && _MSC_VER < 1300) && RETAIL_COMPATIBLE_CRC`
to
`#if USING_STLPORT && RETAIL_COMPATIBLE_CRC`?

Comment (Mauller) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7880]: To me the issue is a compatibility problem between STLPORT and STDLIB with how they use iterators.

So personally i would say the second is more correct to cover the problem.

File: GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp [hunk 2]
```diff
@@ -7877,12 +7877,36 @@ void ScriptEngine::setSequentialTimer(Team *team, Int frameCount)
...
+			SequentialScript *seqScript = (*it);
+			if (seqScript) {
+				DEBUG_LOG(("Sequential script %s appears to be in an infinite loop.\n", 
+					seqScript->m_scriptToExecuteSequentially->getName().str()));
+			}
+			++it;
+			continue;
+		}
+
+		lastIt = it;
```
Comment (xezon) [inline GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7908]: I am trying to understand what exactly is different here to cause retail incompatibility.

greptileai raised a good point:
> The logic difference matters because iterator comparison tracks position while pointer comparison tracks object identity, which can produce different behavior in edge cases where the same script object appears at multiple positions

Is it perhaps possible that the original fix #629 is not right? Can you make another test and instead of using `if ((*it) == lastScript)` use a index based position comparison and then test that against VC6 replay? This would then hint to the AI drawing the correct conclusion that the actual behaviour was changed with the fix when it should not have.

Comment (Caball009) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7908]: If I recall correctly the replays enter the `MAX_SPIN_COUNT` branch for the original EA code but not our code.

Comment (xezon) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7908]: It is suspicious yes.

Comment (Caball009) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7908]: Would be easier to check out if I had a convenient way to reproduce.

Comment (Caball009) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7908]: > Can you make another test and instead of using if ((*it) == lastScript) use a index based position comparison and then test that against VC6 replay?

I can test if you can rewrite the loop.

Comment (xezon) [reply GeneralsMD/Code/GameEngine/Source/GameLogic/ScriptEngine/ScriptEngine.cpp:7908]: Ok. Please try this: #2129

**Comment by xezon** [comment]
I think this can be closed then.

**Comment by Caball009** [comment]
> I think this can be closed then.

I think so too. We'll continue with yours: #2129
